keras/layers/rnn/base_rnn.py

```diff
@@ -625,6 +625,13 @@ def __call__(self, inputs, initial_state=None, constants=None, **kwargs):
                 kwargs["constants"] = constants
             return super().__call__(inputs, **kwargs)
 
+    def _gen_state_names(self):
```

Comment:
Looking at the test output, in case we have a cell that has nested states, the state name will become:

state:0, state_1:0, state_2:0, ...

which is bit weird, how about just use the name state, and let the tf.Variable to increment the count by itself?

Reply:
I'm not sure how to achieve that. For testing I removed the `LSTM._gen_state_names`, and changed the `RNN._gen_state_names` to:

```python
def _gen_state_names(self):
   while True:
       yield "state"
```

And the testcase to:
```python
def test_unique_state_names(self):
        inp = keras.Input(batch_shape=[3, None, 3])
        a = keras.layers.LSTM(units=3, stateful=True)
        b = keras.layers.LSTM(units=3)  # no states
        c = keras.layers.LSTM(units=3, stateful=True)
        _ = keras.Model(inp, [a(inp), b(inp), c(inp)])

        self.assertEqual(a.states[0].name, f"{a.name}/state:0")
        self.assertEqual(a.states[1].name, f"{a.name}/state:1")
        self.assertEqual(b.states[0], None)
        self.assertEqual(b.states[1], None)
        self.assertEqual(c.states[0].name, f"{c.name}/state:0")
        self.assertEqual(c.states[1].name, f"{c.name}/state:1")
```

Which then fails with:
```
in test_unique_state_names
    self.assertEqual(a.states[1].name, f"{a.name}/state:1")
AssertionError:
- lstm/state:0
?            ^
+ lstm/state:1
?
```

To me it seems that `tf.Variable` doesn't increase the `:x` count automatically, but I also don't see any parameter to enforce this.

Reply:
@qlzh727 : Any idea how to achieve this increase of the variable index?